app/server/lib/AttachmentStore.ts

```diff
@@ -0,0 +1,174 @@
+import { joinKeySegments, StreamingExternalStorage } from "app/server/lib/ExternalStorage";
+import * as fse from "fs-extra";
+import * as stream from "node:stream";
+import * as path from "path";
+
+export type DocPoolId = string;
+type FileId = string;
+
+
+// Minimum required info from a document
+// Compatible with Document entity for ease of use
+export interface AttachmentStoreDocInfo {
+  id: string;
+  // This should NOT be an optional: the programmer should make an explicit choice for this value to prevent
+  // accidental omission breaking attachments.
+  trunkId: string | null | undefined;
+}
+
+/**
+ * Gets the correct pool id for a given document, given the document's id and trunk id.
+ *
+ * Attachments are stored in a "Document Pool", which is used to manage the attachments' lifecycle.
+ * Document pools are shared between snapshots and forks, but not between documents. This provides quick forking
+ * and snapshotting (not having to copy attachments), while avoiding more complex systems like reference tracking.
+ *
+ * Generally, the pool id of a document should be its trunk id if available (because it's a fork),
+ * or the document's id (if it isn't a fork).
+ *
+ * This means that attachments need copying to a new pool when a document is copied.
```

Comment:
I don't think we've talked explicitly about the consequences of document copying now being a potentially slow operation. As long as a copy fails cleanly if there's a problem, I think that would be the main thing.

Reply:
I don't think we have either - might be worth a chat? 

I think the copy behaviour is missing from this PR to be honest (:grimacing:). It might be worth opening as a separate PR at this point?